base_layer/core/src/blockchain/chain.rs

```diff
@@ -55,18 +53,44 @@ pub struct Chain {
 }
 
 impl Chain {
-    pub fn new() -> Result<Chain, ChainError> {
-        let chain = Chain {
-            block_chain_state: BlockchainState::new()?,
-            orphans: HashMap::new(),
-            current_total_pow: ProofOfWork::default(),
-        };
-        Ok(chain)
+
+    /// Adds a new block to the persistent state.
+    ///
+    /// This process includes:
+    ///   * Adds the range proof hashes to the range proof MMR
+    ///   * Marks all inputs in the block as spent
+    ///   * Adds all outputs to the UTXO set and MMR
+    ///   * Adds the transaction kernels to the kernel MMR
+    ///   * Adds the block header to the header MMR
+    /// An error is returned if:
+    ///   * the block has already been added (by checking whether the block header hash exists already)
+    ///   * any of the inputs were not in the UTXO set or were marked as spent already
+    ///   * There was a problem creating the checkpoints
+    pub fn add_new_block(&mut self, new_block: &Block) -> Result<(), StateError> {
+        if let Some(_) = self.headers.get_object(&new_block.header.hash()) {
+            return Err(StateError::DuplicateBlock);
+        }
+        // We add the range proofs just to strip them out again 2 lines later???
+        // TODO Clearly there's an optimisation here somewhere
+        for output in &new_block.body.outputs {
+            self.range_proofs.push(output.proof().clone())?;
+        }
+        self.mark_outputs_as_spent(&new_block.body.inputs)?;
+        self.strip_range_proofs(&new_block.body.outputs);
+        // All seems valid, lets add the objects to the state
+        for output in &new_block.body.outputs {
+            self.utxos.push(output.clone().into())?;
+        }
+        self.kernels.append(new_block.body.kernels.clone())?;
+        self.headers.push(new_block.header.clone())?;
+        // Apply a new checkpoint so that we can rewind to this point in the future if necessary
+        self.check_point_state()
     }
 
+
     /// This function will process a newly received block
-    pub fn process_new_block(&mut self, new_block: Block, consensus_rules: &ConsensusRules) -> Result<(), ChainError> {
-        let result = match self.block_chain_state.process_new_block(&new_block, consensus_rules) {
+    pub fn add_new_block(&mut self, new_block: Block, consensus_rules: &ConsensusRules) -> Result<(), ChainError> {
```

Comment:
Is this method still needed? I'm not sure whether this is the main method to call or the one above?

Reply:
This is actually just dangling code right now